r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -76,13 +131,69 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         Ok(res)
     }
 
-    /// Constraints for reducing the result of a multiplication mod p, to get an original representation.
+    /// Reducing a NonNativeMulResultGadget back to a non-native field gadget
+    /// in normal form. Assumes that 
+    /// ``
+    ///     2 * bits_per_limb + surfeit' <= CAPACITY - 2.
+    /// ``
+    /// where 
+    /// ``
+    ///     bits_per_limb = NonNativeFieldParams::bits_per_limb, 
+    ///     surfeit' =  log(num_limbs + num_adds + 1), 
+    ///     num_limbs = NonNativeFieldParams::num_limbs,
+    /// ``
+    /// and `num_adds` is as in the NonNativeMulResultGadget.
+    // Costs
+    // ``
+    //     C =  len(p) + surfeit' 
+    //          + len(p) + num_limbs^2
+    //          +  (num_groups - 1) * (3 + bits_per_limb + surfeit') + 1
+    // ``
+    // constraints, with `num_groups = Ceil(len(p) / S)` and
+    // ``
+    //  S - 1 = Floor[
+    //          (ConstraintF::CAPACITY - 2 - (2 * bits_per_limb + surfeit') / bits_per_limb
+    //      ] 
+    //        = Floor[
+    //          (ConstraintF::CAPACITY - 2 - surfeit') / bits_per_limb
+    //      ] - 2.
+    // ``
     pub fn reduce<CS: ConstraintSystemAbstract<ConstraintF>>(
         &self,
         mut cs: CS,
     ) -> Result<NonNativeFieldGadget<SimulationF, ConstraintF>, SynthesisError> {
+        debug_assert!(
+            self.check(),
+            "reduce(): check() failed on input mul result gadget" 
+        );
+        // This is just paraphrasing the reduction of non-natives. We enforce the large integer 
+        // equality
+        // ``
+        //    Sum_{i=0..} limb[i] * A^i = k * p + r,
+        // ``
+        // where `0 <= r < 2^len(p)` and `k >= 0`, by means of `group_and_check_equality()`. 
+        // As the left hand side is length bounded by
+        // ``
+        //   Sum_{i=0..} limb[i] * A^i < (num_adds + 1) * 2^{2*len(p)}
+        // ``
+        // the quotient `k` is length bounded by
+        // ``
+        //      len(k) <= len(p) + len(num_adds + 1). 
+        // `` 
         let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits());
 
+        // Note: To assure that the limb-wise computation of the `k * p + r` does not 
+        // exceed the capacity bound, we demand 
+        // ``
+        //     2 * bits_per_limb + surfeit'  <= CAPACITY,
+        // ``
+        // with `surfeit' =  log(num_limbs + num_adds + 1)`.
```

Comment:
This is not completely correct in my opinion: it should be len((num_limbs + 2^(len(num_adds + 1)) ). I will explain this more precisely below.

Reply:
I explained below why this is correct.